### PR TITLE
[LiftOff] Update types to v4

### PR DESCRIPTION
**Pull request by crutchcorn**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/gulpjs/liftoff/releases/tag/v4.0.0
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/liftoff/index.d.ts [hunk 1]
```diff
@@ -17,65 +18,58 @@ declare class Liftoff extends EventEmitter {
...
+    prepare(options: Liftoff.PrepareOptions, callback: (env: Liftoff.LiftoffEnv) => void): void;
+    execute(env: Liftoff.LiftoffEnv, callback: (env: Liftoff.LiftoffEnv) => void): void;
 
     addListener(
-        event: 'require',
+        event: 'preload:success',
         listener: (name: string, module: ExtensionDescriptor) => void
     ): this;
-    addListener(event: 'requireFail', listener: (name: string, err: any) => void): this;
+    addListener(event: 'preload:failure' | 'preload:before', listener: (name: string, err: any) => void): this;
```
Comment (BendingBender) [inline types/liftoff/index.d.ts:28]: - The `preload:before` event doesn't pass an `err` parameter.
- The `err` parameter's type can be changed to `Error`

File: types/liftoff/index.d.ts [hunk 2]
```diff
@@ -17,65 +18,58 @@ declare class Liftoff extends EventEmitter {
...
-        opts: Liftoff.LaunchOptions,
-        callback: (this: this, env: Liftoff.LiftoffEnv) => void
-    ): void;
+    prepare(options: Liftoff.PrepareOptions, callback: (env: Liftoff.LiftoffEnv) => void): void;
+    execute(env: Liftoff.LiftoffEnv, callback: (env: Liftoff.LiftoffEnv) => void): void;
 
     addListener(
-        event: 'require',
+        event: 'preload:success',
         listener: (name: string, module: ExtensionDescriptor) => void
```
Comment (BendingBender) [inline types/liftoff/index.d.ts:26]:
```suggestion
        event: 'preload:success',
        listener: (name: string, module: unknown) => void
```

**Review by BendingBender — CHANGES_REQUESTED** [review]
- Please remove the `Extension` import.
- The `forcedFlags` option in the `PrepareOptions` interface was removed.
- The `loader:success` and `loader:failure` events are missing.
- Definition for `EventEmitter.off` for the event `preload:success` is missing.
- We should stop using any imports from `interpret` module, it's not used by this library. The only thing that will remain in use (after you've fixed the review comments below) is the `Extensions` import. You can replace it with: `PathSpec['extensions']`.

**Comment by crutchcorn** [comment]
@BendingBender Thank you so much for the in-depth review!

I've made all requested changes, let me know what else I can change

File: types/liftoff/index.d.ts [hunk 3]
```diff
@@ -17,65 +17,63 @@ declare class Liftoff extends EventEmitter {
...
+    removeAllListeners(event?: 'preload:success' | 'preload:failure' | 'preload:before' | 'respawn'): this;
+    listeners(event: 'preload:success' | 'preload:failure' | 'preload:before' | 'respawn'): Function[]; // tslint:disable-line:ban-types
+    rawListeners(event: 'preload:success' | 'preload:failure' | 'preload:before' | 'respawn'): Function[]; // tslint:disable-line:ban-types
+    emit(event: 'preload:success', name: string, module: unknown): boolean;
+    emit(event: 'preload:failure' | 'preload:before', name: string, err: Error): boolean;
     emit(event: 'respawn', flags: string[], child: NodeJS.Process): boolean;
-    eventNames(): Array<'require' | 'requireFail' | 'respawn'>;
-    listenerCount(type: 'require' | 'requireFail' | 'respawn'): number;
+    eventNames(): Array<'preload:success' | 'preload:failure' | 'preload:before' | 'respawn'>;
+    listenerCount(type: 'preload:success' | 'preload:failure' | 'preload:before' | 'respawn'): number;
```
Comment (BendingBender) [inline types/liftoff/index.d.ts:76]: Ok, I've tried out something new to reduce repetition and add uniformity to definitions. You can replace this whole block with the following definition:
```ts
    addListener<TEvent extends keyof Liftoff.Events>(event: TEvent, listener: Liftoff.Events[TEvent]): this;
    addListener(event: string | symbol, listener: (...args: any[]) => void): this;
    on<TEvent extends keyof Liftoff.Events>(event: TEvent, listener: Liftoff.Events[TEvent]): this;
    on(event: string | symbol, listener: (...args: any[]) => void): this;
    once<TEvent extends keyof Liftoff.Events>(event: TEvent, listener: Liftoff.Events[TEvent]): this;
    once(event: string | symbol, listener: (...args: any[]) => void): this;
    prependListener<TEvent extends keyof Liftoff.Events>(event: TEvent, listener: Liftoff.Events[TEvent]): this;
    prependListener(event: string | symbol, listener: (...args: any[]) => void): this;
    prependOnceListener<TEvent extends keyof Liftoff.Events>(event: TEvent, listener: Liftoff.Events[TEvent]): this;
    prependOnceListener(event: string | symbol, listener: (...args: any[]) => void): this;
    removeListener<TEvent extends keyof Liftoff.Events>(event: TEvent, listener: Liftoff.Events[TEvent]): this;
    removeListener(event: string | symbol, listener: (...args: any[]) => void): this;
    off<TEvent extends keyof Liftoff.Events>(event: TEvent, listener: Liftoff.Events[TEvent]): this;
    off(event: string | symbol, listener: (...args: any[]) => void): this;
    listeners<TEvent extends keyof Liftoff.Events>(event: TEvent): Array<Liftoff.Events[TEvent]>;
    listeners(event: string | symbol): Array<(...args: any[]) => void>;
    rawListeners<TEvent extends keyof Liftoff.Events>(event: TEvent): Array<Liftoff.Events[TEvent]>;
    rawListeners(event: string | symbol): Array<(...args: any[]) => void>;
    emit<TEvent extends keyof Liftoff.Events>(event: TEvent, ...args: Parameters<Liftoff.Events[TEvent]>): boolean;
    emit(event: string | symbol, ...args: any[]): boolean;
```

You'll have to additionally add the new `Events` type to the `Liftoff` namespace below:
```ts
    interface Events {
        'loader:success': (name: string, module: unknown) => void;
        'preload:success': (name: string, module: unknown) => void;
        'loader:failure': (name: string, err: Error) => void;
        'preload:failure': (name: string, err: Error) => void;
        'preload:before': (name: string) => void;
        respawn: (flags: string[], child: NodeJS.Process) => void;
    }
```

File: types/liftoff/liftoff-tests.ts [hunk 4]
```diff
@@ -108,70 +108,51 @@ new Liftoff({
...
 });
-new Liftoff().on('requireFail', (name, err) => {
+new Liftoff().on('loader:failure', (name, err) => {
     // $ExpectType string
     name;
-    // $ExpectType any
+    // $ExpectType Error
     err;
 });
 new Liftoff().on('respawn', (flags, child) => {
```
Comment (BendingBender) [inline types/liftoff/liftoff-tests.ts:158]: You can replace the event tests with the following:
```ts
const liftoff = new Liftoff();

liftoff.addListener('preload:before', name => {
    name; // $ExpectType string
});
liftoff.addListener('preload:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.addListener('preload:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.addListener('loader:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.addListener('loader:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.addListener('respawn', (flags, child) => {
    flags; // $ExpectType string[]
    child; // $ExpectType Process
});
liftoff.addListener('foo', (...args) => {
    args; // $ExpectType any[]
});

liftoff.on('preload:before', name => {
    name; // $ExpectType string
});
liftoff.on('preload:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.on('preload:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.on('loader:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.on('loader:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.on('respawn', (flags, child) => {
    flags; // $ExpectType string[]
    child; // $ExpectType Process
});
liftoff.on('foo', (...args) => {
    args; // $ExpectType any[]
});

liftoff.once('preload:before', name => {
    name; // $ExpectType string
});
liftoff.once('preload:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.once('preload:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.once('loader:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.once('loader:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.once('respawn', (flags, child) => {
    flags; // $ExpectType string[]
    child; // $ExpectType Process
});
liftoff.once('foo', (...args) => {
    args; // $ExpectType any[]
});

liftoff.prependListener('preload:before', name => {
    name; // $ExpectType string
});
liftoff.prependListener('preload:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.prependListener('preload:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.prependListener('loader:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.prependListener('loader:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.prependListener('respawn', (flags, child) => {
    flags; // $ExpectType string[]
    child; // $ExpectType Process
});
liftoff.prependListener('foo', (...args) => {
    args; // $ExpectType any[]
});

liftoff.prependOnceListener('preload:before', name => {
    name; // $ExpectType string
});
liftoff.prependOnceListener('preload:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.prependOnceListener('preload:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.prependOnceListener('loader:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.prependOnceListener('loader:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.prependOnceListener('respawn', (flags, child) => {
    flags; // $ExpectType string[]
    child; // $ExpectType Process
});
liftoff.prependOnceListener('foo', (...args) => {
    args; // $ExpectType any[]
});

liftoff.removeListener('preload:before', name => {
    name; // $ExpectType string
});
liftoff.removeListener('preload:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.removeListener('preload:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.removeListener('loader:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.removeListener('loader:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.removeListener('respawn', (flags, child) => {
    flags; // $ExpectType string[]
    child; // $ExpectType Process
});
liftoff.removeListener('foo', (...args) => {
    args; // $ExpectType any[]
});

liftoff.off('preload:before', name => {
    name; // $ExpectType string
});
liftoff.off('preload:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.off('preload:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.off('loader:success', (name, module) => {
    name; // $ExpectType string
    module; // $ExpectType unknown
});
liftoff.off('loader:failure', (name, err) => {
    name; // $ExpectType string
    err; // $ExpectType Error
});
liftoff.off('respawn', (flags, child) => {
    flags; // $ExpectType string[]
    child; // $ExpectType Process
});
liftoff.off('foo', (...args) => {
    args; // $ExpectType any[]
});

liftoff.listeners('preload:before'); // $ExpectType ((name: string) => void)[]
liftoff.listeners('preload:success'); // $ExpectType ((name: string, module: unknown) => void)[]
liftoff.listeners('preload:failure'); // $ExpectType ((name: string, err: Error) => void)[]
liftoff.listeners('loader:success'); // $ExpectType ((name: string, module: unknown) => void)[]
liftoff.listeners('loader:failure'); // $ExpectType ((name: string, err: Error) => void)[]
liftoff.listeners('respawn'); // $ExpectType ((flags: string[], child: Process) => void)[]
liftoff.listeners('foo'); // $ExpectType ((...args: any[]) => void)[]

liftoff.rawListeners('preload:before'); // $ExpectType ((name: string) => void)[]
liftoff.rawListeners('preload:success'); // $ExpectType ((name: string, module: unknown) => void)[]
liftoff.rawListeners('preload:failure'); // $ExpectType ((name: string, err: Error) => void)[]
liftoff.rawListeners('loader:success'); // $ExpectType ((name: string, module: unknown) => void)[]
liftoff.rawListeners('loader:failure'); // $ExpectType ((name: string, err: Error) => void)[]
liftoff.rawListeners('respawn'); // $ExpectType ((flags: string[], child: Process) => void)[]
liftoff.rawListeners('foo'); // $ExpectType ((...args: any[]) => void)[]

liftoff.emit<'preload:before'>('preload:before', 'foo'); // $ExpectType boolean
liftoff.emit<'preload:success'>('preload:success', 'foo', 'bar'); // $ExpectType boolean
liftoff.emit<'preload:failure'>('preload:failure', 'foo', new Error('foo')); // $ExpectType boolean
liftoff.emit<'loader:success'>('loader:success', 'foo', 'bar'); // $ExpectType boolean
liftoff.emit<'loader:failure'>('loader:failure', 'foo', new Error('foo')); // $ExpectType boolean
liftoff.emit<'respawn'>('respawn', ['foo'], null as any as NodeJS.Process); // $ExpectType boolean
liftoff.emit('foo', 'bar'); // $ExpectType boolean
```

File: types/liftoff/liftoff-tests.ts [hunk 5]
```diff
@@ -108,70 +108,51 @@ new Liftoff({
...
-        },
-    },
-    () => {}
-);
-new Liftoff().launch({ completion: '' }, () => {});
+Hacker.prepare({}, (env) => {
+    Hacker.execute(env, (env) => {
+        // Do post-execute things
+    });
+});
```
Comment (BendingBender) [inline types/liftoff/liftoff-tests.ts:128]: Please replace this test with the more thorough one here:
```ts
Hacker.prepare({}, env => {
    env; // $ExpectType LiftoffEnv
    env.cwd; // $ExpectType string
    env.preload; // $ExpectType string[]
    env.configNameSearch; // $ExpectType string[]
    env.configPath; // $ExpectType string | undefined
    env.configBase; // $ExpectType string | undefined
    env.modulePath; // $ExpectType string | undefined
    env.modulePackage; // $ExpectType { [key: string]: any; }
    env.configFiles; // $ExpectType { [extensions: string]: { [path: string]: string | null; }; }
    env.config; // $ExpectType { [key: string]: any; }
    env.completion; // $ExpectType boolean | undefined

    Hacker.execute(env, function(env, argv) {
        this; // $ExpectType Liftoff
        env; // $ExpectType LiftoffEnv
        argv; // $ExpectType string[]
    });
    Hacker.execute(env, ['--foo'], function(env, argv) {
        this; // $ExpectType Liftoff
        env; // $ExpectType LiftoffEnv
        argv; // $ExpectType string[]
    });
});
```

**Review by BendingBender — CHANGES_REQUESTED** [review]
The `LiftoffEnv` interface has actually changed a litte more:

```ts
    interface LiftoffEnv {
        /**
         * the current working directory
         */
        cwd: string;
        /**
         * an array of modules that liftoff tried to pre-load
         */
        preload: string[];
        /**
         * the config files searched for
         */
        configNameSearch: string[];
        /**
         * the full path to your configuration file (if found)
         */
        configPath: string | undefined;
        /**
         * the base directory of your configuration file (if found)
         */
        configBase: string | undefined;
        /**
         * the full path to the local module your project relies on (if found)
         */
        modulePath: string | undefined;
        /**
         * the contents of the local module's package.json (if found)
         */
        modulePackage: { [key: string]: any };
        /**
         * an object of filepaths for each found config file (filepath values will be null if not found)
         */
        configFiles: { [extensions: string]: { [path: string]: string | null } };
        /**
         * The paths and contents of the found config files.
         */
        config: { [key: string]: any };
        completion: boolean | undefined;
    }
```

Hopefully this is it this time around...

**Comment by crutchcorn** [comment]
@BendingBender thank you so so much for the work you've done on this. I truly apologize for having had made you do so much work here 😅 I didn't anticipate this to be such a lofty lift on your end - apologies!

Changes requested have been made.

**Comment by BendingBender** [comment]
@crutchcorn Don't worry, no need for apologies. It's a complex module to type and it's better to have someone look over the changes. And after all, I helped you by choice, noone forced me to do it.

**Review by BendingBender — APPROVED** [review]

**Comment by crutchcorn** [comment]
Ready to merge

**Pull request merged**